pkg/provider/bitwarden/client.go

```diff
@@ -0,0 +1,311 @@
+	}
+
+	if data.GetSecretKey() == "" {
+		return fmt.Errorf("pushing the whole secret is not yet implemented")
+	}
+
+	if data.GetRemoteKey() == "" {
+		return fmt.Errorf("remote key must be defined")
+	}
+
+	if data.GetProperty() == "" {
+		return fmt.Errorf("property must be defined on secret to determine which project the secret should be pushed to")
+	}
+
+	value, ok := secret.Data[data.GetSecretKey()]
+	if !ok {
+		return fmt.Errorf("failed to find secret key in secret with key: %s", data.GetSecretKey())
+	}
+
+	note, err := utils.FetchValueFromMetadata(NoteMetadataKey, data.GetMetadata(), "")
+	if err != nil {
+		return fmt.Errorf("failed to fetch note from metadata: %w", err)
+	}
+
+	// ListAll Secrets for an organization. If the key matches our key, we GetSecret that and do a compare.
+	remoteSecrets, err := p.bitwardenSdkClient.ListSecrets(ctx, spec.Provider.BitwardenSecretsManager.OrganizationID)
+	if err != nil {
+		return fmt.Errorf("failed to get all secrets: %w", err)
+	}
+
+	for _, d := range remoteSecrets.Data {
+		if d.Key != data.GetRemoteKey() {
+			continue
+		}
+
+		sec, err := p.bitwardenSdkClient.GetSecret(ctx, d.ID)
+		if err != nil {
+			return fmt.Errorf("failed to get secret: %w", err)
+		}
+
+		// If all pushed data matches, we won't push this secret.
+		if sec.Key == data.GetRemoteKey() &&
+			sec.Value == string(value) &&
+			sec.Note == note &&
+			sec.ProjectID != nil &&
+			*sec.ProjectID == data.GetProperty() {
+			// we have a complete match, skip pushing.
+			return nil
+		} else if sec.Key == data.GetRemoteKey() &&
+			sec.Value != string(value) &&
+			sec.Note == note &&
+			sec.ProjectID != nil &&
+			*sec.ProjectID == data.GetProperty() {
+			// only the value is different, update the existing secret.
+			_, err = p.bitwardenSdkClient.UpdateSecret(ctx, SecretPutRequest{
+				ID:             sec.ID,
+				Key:            data.GetRemoteKey(),
+				Note:           note,
+				OrganizationID: spec.Provider.BitwardenSecretsManager.OrganizationID,
+				ProjectIDS:     []string{data.GetProperty()},
+				Value:          string(value),
+			})
+
+			return err
+		}
+	}
+
+	// no matching secret found, let's create it
+	_, err = p.bitwardenSdkClient.CreateSecret(ctx, SecretCreateRequest{
+		Key:            data.GetRemoteKey(),
+		Note:           note,
+		OrganizationID: spec.Provider.BitwardenSecretsManager.OrganizationID,
+		ProjectIDS:     []string{data.GetProperty()},
+		Value:          string(value),
+	})
+
+	return err
+}
+
+// GetSecret returns a single secret from the provider.
+func (p *Provider) GetSecret(ctx context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	if strfmt.IsUUID(ref.Key) {
+		resp, err := p.bitwardenSdkClient.GetSecret(ctx, ref.Key)
+		if err != nil {
+			return nil, fmt.Errorf("error getting secret: %w", err)
+		}
+
+		return []byte(resp.Value), nil
+	}
+
+	if ref.Property == "" {
+		return nil, fmt.Errorf("property must be defined if lookup by name")
+	}
```

Comment:
Generally about the design:

The `ref.Key` is supposed to point to the key, `ref.Property` should point to a property if the secret value is structured data like JSON. Some providers use `gjson` under the hood to extract a nested value (and support gjson syntax in the `ref.Property` field).

Using the `property` to define the projectID is problematic, because it is used for a different purpose looking at other providers.
I'd suggest to define a schema for the `ref.Key`. e.g. the Azure KeyVault provider supports different object types:
1. `secret/{name}`
2. `certificate/{name}`
3. `key/{name}`

In AWS Secrets Manager you can define secret versions with UUID or by label
1. `uuid/{...}` to fetch a specific secret version by uuid
2. `AWSCURRENT` to fetch the "current" version of a secret (this is a well-known AWS SM label)

For Bitwarden, we can also introduce a schema to specify the secret key. I didn't look to deep into it, this is my understanding:
1.  `123e4567-e89b-12d3-a456-426614174000` attempt to parse uuid, then the secret is unique across all organizations/projects/$world
2. `123e4567-e89b-12d3-a456-426614174000/{name}` attempt to parse the first segment as project id, then the next (+remaining?) segment(s) as secret name?

We could be even more explicit by making the org mandatory and hoist it to the SecretStore provider spec, To me it seems this resembles a "tenant" in the bitwarden world. If we hoist it we don't need the schema above and instead define either the uuid or secret name, no need for a project id then.

Reply:
The organisation id _is_ in the store ref. 🤔😀

That's a good idea to have a structured key! I'll work on that. Thanks. 🤩

Reply:
>123e4567-e89b-12d3-a456-426614174000 attempt to parse uuid, then the secret is unique across all organizations/projects/$world

We already do this 👍 

>123e4567-e89b-12d3-a456-426614174000/{name} attempt to parse the first segment as project id, then the next (+remaining?) segment(s) as secret name?

Yep, that makes sense. 👍  Thank you for the suggestion! I knew I would mix up the property somehow. :D Incredible that after all this time, I still confuse key/property and remotekey. :D

Reply:
>We could be even more explicit by making the org mandatory and hoist it to the SecretStore provider spec, To me it seems this resembles a "tenant" in the bitwarden world. If we hoist it we don't need the schema above and instead define either the uuid or secret name, no need for a project id then.

So, the org is in the store. The project id is needed in case of names because the name is not unique ever. So we add the projectID to be sure what we are changing is the thing we are changing. And if there is more than one name in the same project we will error out. Otherwise we could risk modifying the wrong secret.

Therefor if the Key is not a UUID we will always need {projectID}/{name} format.

Reply:
I meant that maybe we can add the project id to the store as well, it seems this is supposed to be a unit of separation (a "tenant") - then we wouldn't need to have a complex ref.key pattern.

Reply:
Ah that makes more sense. 🤣🤣

Reply:
Let me think about this and see what others think because my impression is that you can have a LOT of projects in an organization. That would mean that you would have a proliferation of secret stores potentially for many projects.

But then again maybe you're right. Because you have an ORG for an organisation then a prod and dev project or a project per team. Then it would be unwieldy to use a single store and then prefix with project. 

Hmmmmmmmmmmmmmmm 🧐

Reply:
Okay, done. Let's see how this holds up. :)